namespace/cache_resolver.go

```diff
@@ -0,0 +1,123 @@
+package namespace
+
+import (
+	"container/list"
+	"sync"
+	"time"
+)
+
+const DefaultMaxCachedEntries = 512
+
+type cacheEntry struct {
+	name    string
+	created time.Time
+	entries *Entries
+}
+
+func newCacheEntry(name string, entries *Entries) cacheEntry {
+	return cacheEntry{name, time.Now(), entries}
+}
+
+type entriesCache struct {
+	mutex       sync.Mutex
+	cache       map[string]cacheEntry
+	expireAfter time.Duration
+	maxEntries  int
+	/* Contains pointers to cache entries sorted by the time of their addition.
+	 * Entry added last will be at the end. */
+	expirationQueue *list.List
+}
+
+func newEntriesCache(expireAfter time.Duration, maxEntries int) *entriesCache {
+	var expirationQueue *list.List
+	if maxEntries > 0 {
+		expirationQueue = list.New()
+	}
+	return &entriesCache{
+		cache:           make(map[string]cacheEntry),
+		expireAfter:     expireAfter,
+		maxEntries:      maxEntries,
+		expirationQueue: expirationQueue,
+	}
+}
+
+// Must only be called from lookup method.
+func (sc *entriesCache) garbageCollectExpired() {
+	if sc.expirationQueue == nil {
+		return
+	}
+	now := time.Now()
+	elem := sc.expirationQueue.Front()
+	for elem != nil && elem.Value.(*cacheEntry).created.Add(sc.expireAfter).Before(now) {
+		delete(sc.cache, elem.Value.(*cacheEntry).name)
+		next := elem.Next()
+		sc.expirationQueue.Remove(elem)
+		elem = next
+	}
+}
+
+func (sc *entriesCache) lookup(name string) *Entries {
+	sc.mutex.Lock()
+	defer sc.mutex.Unlock()
+	entry, exists := sc.cache[name]
+	if exists {
+		if sc.expireAfter == 0 || !entry.created.Add(sc.expireAfter).Before(time.Now()) {
+			return entry.entries
+		}
+		sc.garbageCollectExpired()
```

Comment:
I haven't considered it until now. The LRU approach IMHO isn't appropriate for caching http responses. It would prevent the most frequently searched items from ever being removed. If a configuration change occured at discovery endpoint, Docker daemon would stay oblivious to that for quite a while - until restarted. That's what this expiring cache is trying to solve.